webgpu/generator.py

```diff
@@ -27,79 +28,110 @@
+    result = ""
+
+    for entrypoint in entrypoints:
+        result += "<entrypoint> = \"{}\"\n".format(entrypoint)    
 
-def generate_lines(grammar, num_lines):
-    output = ''
-    output += grammar._generate_code(num_lines)
+    return result
 
-    return output
+def parse_bindings(shaders):
+    bindings = []
 
-def GenerateNewSample(template, jsgrammar, wgslgrammar):
-    """Parses grammar rules from string.
+    for shader in shaders:
+        tokens = shader.split()
+        for i, token in enumerate(tokens):
+            if "@binding" in token:
+                bindings.append(token[token.index("@binding(") + len("@binding("):-1])
 
-    Args:
-      template: A template string.
-      jsgrammar: Grammar for generating JS code.
-      wgslgrammar: Grammar for generating WGSL shader code.
+    result = ""
+    for binding in bindings:
+        result += "<BindInt> = {}\n".format(binding)
 
-    Returns:
-      A string containing sample data.
-    """
+    return result
 
-    result = template
-    handlers = False
+def generate_function_body(webgpugrammar, num_lines):
+    js = ''
+    js += webgpugrammar._generate_code(num_lines)
 
-    # WebGPU JS
+    return js
+
+def generate_new_sample(template, webgpugrammar):
+    result = template
     while '<webgpufuzz>' in result:
-        numlines = _N_MAIN_LINES
-        if handlers:
-            numlines = _N_EVENTHANDLER_LINES
-        else:
-            handlers = True
         result = result.replace(
             '<webgpufuzz>',
-            generate_lines(jsgrammar, numlines),
+            generate_function_body(webgpugrammar, _N_MAIN_LINES),
             1
         )
 
     return result
 
 
-def generate_samples(grammar_dir, outfiles):
-    """Generates a set of samples and writes them to the output files.
+def generate_samples(template, grammar_dir, outfiles):
+    extra = ""
+    shaders_dir = os.path.join(grammar_dir, "wgsl/*.wgsl")
+    shader_files = glob.glob(shaders_dir)
 
-    Args:
-      grammar_dir: directory to load grammar files from.
-      outfiles: A list of output filenames.
-    """
+    shaders = []
+    for i in range(_N_SHADERS):
+        shader_path = random.choice(shader_files)
+        with open(shader_path) as fp:
+            shader_src = fp.read()
+            shaders.append(shader_src)
+    
+    with open(os.path.join(grammar_dir, template), "r") as fp:
+        template_contents = fp.read()
 
-    f = open(os.path.join(grammar_dir, 'template.html'))
-    template = f.read()
-    f.close()
+    SHADER_CONST = "<shader%s>"
+    for i, shader in enumerate(shaders[:-1]):
+        template_contents = template_contents.replace(SHADER_CONST % str(i), shader)
 
-    jsgrammar = Grammar()
-    err = jsgrammar.parse_from_file(os.path.join(grammar_dir, 'webgpu.txt'))
-    if err > 0:
-        print('There were errors parsing WebGPU JS grammar')
-        return
+    extra += parse_entrypoints(shaders)
+    extra += parse_bindings(shaders)
```

Comment:
I wonder if it would make sense to pass these extras as initial_variables params to `_generate_code` (see `generator.py` in parent directory for example), rather than appending grammar definition.

Reply:
I think this is a good option and I'm happy to add it. I will make one argument for injecting them into the grammar definition, and it's that it reflects the probabilities from the sampled wgsl files.